### PR TITLE
C# Facebook Events Sample

### PR DESCRIPTION
**Pull request by carlosscastro**

### CODE REVIEW

File: csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.csproj [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    <AssemblyName>Facebook_Events_Bot</AssemblyName>
+    <RootNamespace>Facebook_Events_Bot</RootNamespace>
+    <CodeAnalysisRuleSet>Facebook-Events-Bot.ruleset</CodeAnalysisRuleSet>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.7" />
+    <PackageReference Include="Microsoft.Bot.Builder.Ai.LUIS" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Version="4.0.0.38762" />
```
Comment (daveta) [inline csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.csproj]: Think you may want to remove these..

File: csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.csproj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    <PackageReference Include="Microsoft.Bot.Builder" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Configuration" Version="4.0.0.38762" />
+    <PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Features" Version="2.9.0" />
```
Comment (daveta) [inline csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.csproj]: >CodeAnalysis [](start = 41, length = 12)

We've been removing this one as well..

File: csharp_dotnetcore/81.Facebook-Events/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        /// </summary>
+        /// <param name="env">Provides information about the web hosting environment an application is running in.</param>
+        /// <seealso cref="https://docs.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-2.1"/>
+        public Startup(IHostingEnvironment env)
+        {
+            var builder = new ConfigurationBuilder()
+                .SetBasePath(env.ContentRootPath)
+                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
+                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
+                .AddEnvironmentVariables();
```
Comment (daveta) [inline csharp_dotnetcore/81.Facebook-Events/Startup.cs]: We've reduced dependency on appsettings, you can remove this.

File: csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.bot [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    "description": "",
+    "secretKey": "",
+    "services": [
+      {
+        "type": "endpoint",
+        "name": "facebook-events",
+        "id": "https://2137f71a.ngrok.io/api/messages",
+        "appId": "1999c181-69b1-4705-a51d-cd3071e8532d",
+        "appPassword": "mgmgkKKY7664!lpSMMN2:]_",
+        "endpoint": "https://2137f71a.ngrok.io/api/messages"
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/Facebook-Events-Bot.bot]: Remove the AppID and Password. 

... and now that they're in Git History, rotate the keys.

File: csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+using Microsoft.Bot.Builder.AI.Luis;
+using Microsoft.Bot.Builder.AI.QnA;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Bot.Builder.Dialogs.Choices;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Facebook_Events_Bot
+{
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs]: Use our standard namespace. Microsoft.BotBuilderSamples

File: csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        private const string DialogId = "question";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SimplePromptBot"/> class.
+        /// </summary>
+        /// <param name="accessors">The state accessors this instance will be needing at runtime.</param>
+        public FacebookEventsBot(BotAccessors accessors)
+        {
+            _dialogs = new DialogSet(accessors.ConversationDialogState);
+            _dialogs.Add(new AttachmentPrompt(DialogId));
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs]: Null check.

File: csharp_dotnetcore/81.Facebook-Events/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
+﻿
+
+# Concepts introduced in this sample
+
+This sample shows how to integrate consume Facebook specific payloads, such as PostBacks, quick replies and optin events. 
+Since Bot Framework supports multiple Facebook pages for a single bot, we also show how to know the page to which the message was sent, so developers can have custom behavior per page.
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/README.md]: integrate and consume

File: csharp_dotnetcore/81.Facebook-Events/README.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+```bash
+git clone https://github.com/Microsoft/botbuilder-samples.git
+```
+
+## Visual Studio
+- Navigate to the samples folder (`BotBuilder-Samples\csharp_dotnetcore\81.Facebook-Events`) and open Facebook-Events-Bot.csproj in Visual Studio 
+- Hit F5
+
+## Visual Studio code
+- Open `BotBuilder-Samples\csharp_dotnetcore\81.Facebook-Events` sample folder.
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/README.md]: casing. code => Code

File: csharp_dotnetcore/81.Facebook-Events/README.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+- Bring up a terminal, navigate to BotBuilder-Samples\81.Facebook-Events folder
+- type 'dotnet run'
+
+## Testing the bot using Bot Framework Emulator
+[Microsoft Bot Framework Emulator](https://github.com/microsoft/botframework-emulator) is a desktop application that allows bot developers to test and debug their bots on localhost or running remotely through a tunnel.
+
+- Install the Bot Framework emulator from [here](https://aka.ms/botframeworkemulator).
+
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/README.md]: casing. emulator => Emulator

File: csharp_dotnetcore/81.Facebook-Events/README.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+### Connect to bot using Bot Framework Emulator **V4**
+- Launch Bot Framework Emulator
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/README.md]: remove "**" around V4

File: csharp_dotnetcore/81.Facebook-Events/README.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+To publish your bot to Azure, follow the [Azure Bot Service documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-deploy-azure?view=azure-bot-service-3.0) which provides simple steps to get your bot in the cloud in a matter of minutes.
+
+### Enable Facebook Channel
+
+The final step to test Facebook-specific features is to publish your bot for the Facebook channel. The Bot Framework makes this very easy, and the detailed steps are explained in the [Bot Framework Channel Documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-connect-facebook?view=azure-bot-service-3.0).
+
+# Further reading
+
+- [Azure Bot Service Introduction](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-overview-introduction?view=azure-bot-service-4.0)
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/README.md]: remove "Introduction"

File: csharp_dotnetcore/81.Facebook-Events/FacebookModel/FacebookMessage.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/FacebookModel/FacebookMessage.cs]: Remove & Sort Using.

File: csharp_dotnetcore/81.Facebook-Events/FacebookModel/FacebookMessage.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
+﻿using Newtonsoft.Json;
+using System;
```
Comment (cleemullins) [inline csharp_dotnetcore/81.Facebook-Events/FacebookModel/FacebookMessage.cs]: // Copyright

**Comment by cleemullins** [comment]
> using Newtonsoft.Json;

Copyright headers on all these file.s 

---
Refers to: csharp_dotnetcore/81.Facebook-Events/FacebookModel/FacebookOptin.cs:1 in de67d1a. [](commit_id = de67d1aaa10ee97bfdae00dd796a8b1cf6bc946d, deletion_comment = False)

File: csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs (hunk 5 above)
Comment (carlosscastro) [reply csharp_dotnetcore/81.Facebook-Events/FacebookEventsBot.cs]: @cleemullins I actually see a lot of samples with namespaces like this. This should be one of the things to do for the final pass on samples.

What is the correct namespace out of the following 2:

```cs
namespace Microsoft.BotBuilderSamples.Facebook_Events_Bot
```
or

```cs
namespace Microsoft.BotBuilderSamples
```

File: csharp_dotnetcore/81.Facebook-Events/README.md (hunk 7 above)
Comment (carlosscastro) [reply csharp_dotnetcore/81.Facebook-Events/README.md]: Fixed the integrate and consume part. optin is listed like this in facebook docs, so I'll leave it as-is for easy text search. Let me know if you have another proposal, happy to reconsider.

Comment (carlosscastro) [reply csharp_dotnetcore/81.Facebook-Events/README.md]: Also changed PostBacks to postbacks to follow facebook docs here.

**Pull request merged**